v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/transforms/SpannerChangeStreamsToOrderedKV.java

```diff
@@ -0,0 +1,267 @@
+                      record.getPartitionToken(),
+                      record.getCommitTimestamp(),
+                      record.getServerTransactionId(),
+                      record.isLastRecordInTransactionInPartition(),
+                      record.getRecordSequence(),
+                      record.getTableName(),
+                      record.getRowType(),
+                      Collections.singletonList(mod),
+                      record.getModType(),
+                      record.getValueCaptureType(),
+                      record.getNumberOfRecordsInTransaction(),
+                      record.getNumberOfPartitionsInTransaction(),
+                      record.getTransactionTag(),
+                      record.isSystemTransaction(),
+                      record.getMetadata()))
+          .forEach(outputReceiver::output);
+    }
+  }
+
+  private static class KeyByIdFn extends DoFn<DataChangeRecord, KV<String, DataChangeRecord>> {
+    private int numberOfBuckets = 1000;
+
+    private KeyByIdFn(int numberOfBuckets) {
+      this.numberOfBuckets = numberOfBuckets;
+    }
+
+    // numberOfBuckets is configured by the user (via orderingPartitionBucketCount parameter)
+    // to match their key cardinality
+    // Here, we are choosing to hash the partition (Spanner primary keys) to a bucket index, in
+    // order to have a deterministic number
+    // of states and timers for performance purposes.
+    // Note that having too many buckets might have undesirable effects if it
+    // results in a low number of records per bucket
+    // On the other hand, having too few buckets might also be problematic, since
+    // many keys will be contained within them.
+
+    @ProcessElement
+    public void processElement(
+        @Element DataChangeRecord record,
+        OutputReceiver<KV<String, DataChangeRecord>> outputReceiver) {
+      // TODO: generate hashcode dynamically based on orderingPartitionKey()
+      int hashCode = (int) record.getMods().get(0).getKeysJson().hashCode();
+      // Hash the received keys into a bucket in order to have a
+      // deterministic number of buffers and timers.
+      String bucketIndex = String.valueOf(hashCode % numberOfBuckets);
+
+      outputReceiver.output(KV.of(bucketIndex, record));
+    }
+  }
+
+  private static class BufferKeyUntilOutputTimestamp
+      extends DoFn<KV<String, DataChangeRecord>, KV<String, Iterable<DataChangeRecord>>> {
+    private static final Logger LOG = LoggerFactory.getLogger(BufferKeyUntilOutputTimestamp.class);
+
+    private long incrementIntervalInSeconds = 2;
+
+    private BufferKeyUntilOutputTimestamp(long incrementIntervalInSeconds) {
+      this.incrementIntervalInSeconds = incrementIntervalInSeconds;
+    }
+
+    @SuppressWarnings("unused")
+    @TimerId("timer")
+    private final TimerSpec timerSpec = TimerSpecs.timer(TimeDomain.EVENT_TIME);
+
+    @StateId("buffer")
+    private final StateSpec<BagState<DataChangeRecord>> buffer = StateSpecs.bag();
```

Comment:
BagState has a 2GB limit. If this pipeline is run to handle backfill data - example to populate 1 day worth of change stream backlog to PubSub - this will fail and enter into a irrecoverable state.

Reply:
This template has some configurable params like `orderingPartitionBucketCount`, `bufferTimerInterval` ,  which has to be configured by user based on their scale. 

`orderingPartitionBucketCount` will limit number of states and timers, `bufferTimerInterval` should help in limiting individual state buffer size. 

Idea is to clearly covey the user that these kinds of issues may occur if the params are not configured/adjusted properly for their scale.